### ISSUE TITLE
Exposing Fields for Selects and supporting OR

### ISSUE THREAD

**Issue by kazetsukaimiko** [ojplg/hrorm#8]
I was thinking about how to go about doing this, and there are a couple of major architectural concerns I'd like to discuss upfront, the first being the most "controversial":

**BUILDER**
It seems like Column should have its field type as a generic, but increasing the number of generics is going to add quite a bit of cognitive complexity to the internals. I also find it awkward how the internals have two types to worry about for representing an entity, and the getter/setter method references on a potential Field class can only ever apply only in one direction. 

It would be one thing if the immutability extended to the Dao as well, but as it is, even if the POJO is immutable hrorm still has update methods exposed for this use case. In test code at least, there's also an assumed builder, or at the very least, a mutable version of the entity class. What about immutables where there's no builder, constructor only?

I think removing the BUILDER types internally and shiftng the ResultSet => Object mapping to later in the pipeline will get you the same functionality with more concise generic typing throughout, and less code. It will even allow fairly easy support for immutables with no mutable sibling/builder class, and potentially, individual final fields on otherwise mutable entities.

Needless to say, I'll never propose changes that introduce dependencies or reflection.

**Template Queries**
Personally I think template selects should go away once Field selects are introduced and work well. I think the template model is great for update(), but since we're both thinking it would be kind of nice if we could make queries more intuitive and powerful, I think we'll end up replacing the Predicate side of SQL generation with some kind of object representing the conditionals internally, and that Predicate model will completely eclipse existing capabilities anyway.

You expressed interest in making the library difficult to misuse- this is an area where there's no contract between your Dao definition and the query input. update can be changed to easily validate against the Dao model:
`update(T template, Field<T, ?>... fields) // Mixin some kind of Operator`

Selects, not so much. Perhaps we can throw exceptions earlier but compile-time safety with them isn't going to happen. 

**Operator**
I mentioned before the current Operator scope creep. I think Operator should be reverted to something like that of an enum, with subclasses to help enforce type and query safety. Values should not be stored on these, ideally. 

**SQL Builder**
I think SQL Builder's responsibilities should be broken up into more distinct parts, Commands (Selections, Updates, Deletes), Predicates (where .... ), and  Modifiers (ORDER BY, LIMIT, DESC, etc). Combined these represent a Query. I know, we cannot support all of SQL. SQL is "big", I agree. I think we can support more of SQL without vastly expanding the internals or interfaces.

Anyway, these are some of the objectives I would have with designing Field support, and making queries easier and more intuitive. Please let me know what, if anything I can do to make these work well with your vision for the library.

**Comment by ojplg** [comment]
There's a lot here. But I thought I would reply to one thing immediately, even at the risk of saying something stupid, which is high.

I am actually coming around to more of your way of thinking about building where clauses.

I am against trying to rebuild SQL in Java objects, but we can perhaps find a middle ground.

I was thinking of something that looked like this:

`Dao<Foo> dao = // well somehow or other we have this.`
`List<Foo> selected = dao.selectWhere("number_column_a", Operator.EQUALS,13L)`
`.and("date_column",Operator.GREATER_THAN, someTime)`
`.and("date_column",Operator.LESS_THAN_OR_EQUAL,otherTIme)`
`.execute();`

I even started a branch (select_experiment) to try to work something up in this direction.

If that can be the model, then we can use a very similar syntax for updates and deletes.

It solves several issues.

I want to digest more of your comments, and I will likely respond to them piecemeal.

**Comment by kazetsukaimiko** [comment]
Not a big deal, we're going to have to hash this out a bit. I want to clarify, that I do not intend to make someone do :

```
Dao<Foo> dao;

SQLObject someRepresentationOfSql = new SQLObject();
someRepresentationOfSql.command(/**/)
someRepresentationOfSql.predicate(/**/)
someRepresentationOfSql.modifier(/**/)

dao.use(someRepresentationOfSql);
```
That is extremely ugly and not my intent. Whatever SQLObject is, I think needs to be as unobtrusive as possible. At the very least, daos _must_ maintain that simple one-line call chain capability for 100% of the functionality offered.

One area where I propose compromise is or/and, as chaining everything serially eventually makes no sense at all. A static method import makes full and/or possible with very little pain:

```
// Did you want to keep everything in the same package?
import static org.hrorm.query.Composite.and;
import static org.hrorm.query.Composite.or;

/**/

LocalDateTime firstStart = LocalDateTime.now().minus(Duration.of(3, ChronoUnit.DAYS));
LocalDateTime firstEnd = LocalDateTime.now();

LocalDateTime secondStart = LocalDateTime.now().minus(Duration.of(30, ChronoUnit.DAYS));
LocalDateTime secondEnd = LocalDateTime.now().minus(Duration.of(25, ChronoUnit.DAYS));

Dao<Foo> dao;
List<Foo> foos = dao.select(
  or(
    // Operator and friends get completely hidden behind the scenes with
    // the right methods on Field<ENTITY, TYPE>.
    Foo.DATE.isAfter(firstStart)
          .and(Foo.DATE.isBefore(firstEnd))
          .and(Foo.TYPE.equalTo(0))
    ,
    Foo.DATE.isAfter(secondStart)
          .and(Foo.DATE.isBefore(secondEnd))
          .and(Foo.TYPE.equalTo(1))
  )
);
```

For field selects, is the direction I was thinking of going. Do we want to maintain non-typesafe/column string name selects as well?

**Comment by kazetsukaimiko** [comment]
To further clarify, "SQLObject" is vastly overstating what I'm proposing. I'm largely thinking a Condition class (extends PredicateClause, is a bottle for Field, Operator, Value), a Composite class (extends PredicateClause, List<Condition> + AND/OR), and a Modifier class (ORDER BY, LIMIT, etc), a PredicateClause superclass that has abstract methods to generate the SQL string for that fragment and another abstract method to create the ordered List of Field to determine where each ? preparedStatement substitution happens.

Field.{equalTo,isBefore,isAfter}(value) all return new Condition(this, operator, value).
Condition.and(condition2) returns new Composite(AND/OR, PredicateClause...)
static Composite.and/or does the same thing.
non-static Composite.and/or appends PredicateClauses to its internal list if it is the same type (and/or). Otherwise embeds another Composite.

I'll look at your branch if it is visible.

**Comment by kazetsukaimiko** [comment]
Oh wow, you WhereClauseBuilder is a very close approximation.

**Comment by ojplg** [comment]
Some random thoughts about ENTITY/BUILDER and immutability.

1. The choice to create an immutable code level Java model has very little to do with whether or not the values of entities can change. So, supporting update() methods for immutable objects makes sense.
2. Immutable classes without builders are almost impossible for frameworks to support transparently. The Java type system (and really, any type system) just has no way to express "here I need a function from a bunch of types of unknown length to an object". I don't know how to support that in hrorm. Builders are good and useful though. That's why they're built into lombok.
3. The distinction between `BUILDER` and `ENTITY` does add considerable complexity to hrorm internals, but it comes with a major win for the client.
4. I am not against having columns advertise the type they support through generics. At various times in hrorm's history, that was part of it. But at the moment, it does not buy me anything. The `Column` interface just has no need of the underlying type. If that changes, well, that changes. `Column` is not part of the exported API of hrorm, at least not now.

**Comment by kazetsukaimiko** [comment]
2.  I think we can go from ResultSet -> some intermediary, which we can then construct the immutable using Function<Intermediary, Immutable>. Again, there's already an assumed mutable version of the immutable.

3. I think we can manage the client win without the internal distinction (and added complexity) living where it does, that's my whole point. 

4. I've been thinking about this, Column and Field are two different concepts. Perhaps there needs to be something like a Mapping<Column, Field> link to keep their functionalities distinct.

I've updated my mock-up of a fluent query example:
https://github.com/kazetsukaimiko/hrorm/tree/fluent-query/src/main/java/org/hrorm/query

Nothing is working yet, but current usage:
https://github.com/kazetsukaimiko/hrorm/blob/fluent-query/src/test/java/org/hrorm/FluentTest.java

**Comment by ojplg** [comment]
So, I did the best I could in terms of keeping the internals of hrorm simple. If they can be made simpler, I am all for it. I just don't see how myself.

If we need to distinguish fields and columns, and that helps somehow, that's no problem for me. If we can use the BUILDER type less often, I am fine with that too.

I care about the functionality provided to the library user and the simplicity of that interface above anything else.

**Comment by kazetsukaimiko** [comment]
Ugh, I've made a mess of my branch exploring this stuff.

Part of the problem is that DaoBuilders are all implementations, there isn't a well-defined interface for them anywhere. I started there, because the number of DaoBuilders out there to choose from makes it ugly. In general, I think classes are doing a little too much and need to be divided up.

Defining daobuilder interfaces has turned this into more or less a complete refactor of the building process, but I think it might be worth it to achieve a single-builder pattern. What do you think? Right now the user has to new up the implementation, personally I don't like that. I want the daobuilder interface to guide the user to the correct implementation.

In order not to make a huge mess, and to keep me from stepping on your toes, I've put all these new classes under a subpackage so they can be fleshed out in parallel with existing builders. I've yet to commit any of this, again its still a mess, I have to revert more or less everything I've done in the main package so I'm just going to separate this out into its own new experimental branch.

Overall I think rather than shoehorning Fields into the old classes it would be better to redesign with smaller classes all around and address the concerns directly.

Oh, and I'm thinking about baking BUILDER into the internals of the new stuff at this point regardless- I think the way you did it may afford some additional flexibility I didn't consider earlier. If I divide up concerns well enough, the additional type isn't going to cost anything.

TL;DR: This is hard. 

Give me some time on this one.

**Comment by ojplg** [comment]
Yeah, I've been trying to warn you that I think this is much harder than you have been letting on.

I thought for a long time and tried many ideas when I originally created the two largely independent DAO builder classes and now there are three.

The problem is that you need to return exactly the correct concrete type from each `with` method, or the type checker will not allow it. It cannot be an interface. It cannot be an abstract class. The methods actually have different signatures, and must to be clean for the client.

This is what I meant in my very first comment to the first ticket you opened.

Similarly for the SQL building. Once you open it up, you find it's a real spider's web of complications. And it's very hard to move anything a small amount without twisting everything else out of shape, or creating a pretty ugly set of special purpose interfaces that only fit together one way and are therefore hard to use, or both.

Be sure you want to spend your time this way. It might not lead to anything good.

**Comment by kazetsukaimiko** [comment]
I had to let this one cook for a few days in the background... I think I have it figured out, but its still a significant effort with questionable return. Starting a new branch with the interfaces to give you an idea where I think this goes.

I'd like to start addressing some of the items mentioned in the README as well, but I think I may have another bug to deal with first.

**Comment by kazetsukaimiko** [comment]
Hey, I've still been chipping away at this. Almost written another response two or three times now feeling this out.
If we want to support building WHERE fluently and consolidate the implementations to be a little less one-off, I think we could do well by reinventing the interfaces first. I'm going to open a PR with _just_ new interfaces for the Daos and Dao builders under a separate subpackage. Once we nail that down I think I can work to refactor the necessary internals.

I went down this path alone a few times already and I was finding myself forced to make decisions that really you should have input on. Interface design is one of those.

For instance, BUILDER: Eliminating it would mean at the very least, that the user would have to define a bidirectional conversion for ENTITY<->BUILDER. A compromise (not one I like- I've warmed up to the BUILDER generic actually), but also a design decision. Another compromise would be to make an "InternalDao" that returns a `Stream<SomethingRepresentingARow> `and then mapping that generic `Something` to BUILDER on the way out- this would open up doors to non-SQL datasources- think SQLDao, etc but more importantly separate SQL concerns and entity mapping concerns- the tradeoff here being of course, you have to isolate all SQL-specific concerns to that layer of the stack.

I'm going to have two subpackages:
org.hrorm.refresh (with the new interfaces)
org.hrorm.refresh.concept (showing how we can make an AnimalBuilder<T> that returns a Cat<T>, Dog<T>, etc, along with their various implementations)

Hopefully staring at the interfaces will make the goals more clear and the work easier.

**Comment by ojplg** [comment]
Take a good look at my branch.

https://github.com/ojplg/hrorm/tree/select_experiment

It's not perfect, but the interface is pretty clean, and it is actually working.

Thinking about interfaces is great, but there's something about working code.

Are you OK with the count stuff for now? Or should we push forward on that before this possibly deep dive?

**Comment by ojplg** [comment]
One other thing that you have touched on: once you have where clause building in place, it should work for updates and deletes as well. Less common than select, but a nice thing to convince yourself you've got the design right.

**Comment by kazetsukaimiko** [comment]
> https://github.com/ojplg/hrorm/tree/select_experiment
> 
For the amount of code, what it delivers is pretty good.

> It's not perfect, but the interface is pretty clean, and it is actually working.
> Thinking about interfaces is great, but there's something about working code.
> 

Yeah, can't beat that.

> Are you OK with the count stuff for now? Or should we push forward on that before this possibly deep dive?

I had to write JDBC code to handle the counts already since it wasn't in HRORM. Unfortunately, that's likely immortal at this point, unless the count and query pieces of work go through pretty quickly.

> One other thing that you have touched on: once you have where clause building in place, it should work for updates and deletes as well.

It will work with counts too, so there's some synergy there.

Anything I can do to help you get the count functionality + fluent selects over the line, I'll focus on. I'm going through your counts branch right now and will add my feedback to the issue.

**Comment by ojplg** [comment]
I updated the select_experiment branch. I took some ideas you had floated, but changed them around a bit. I have something that looks like a column, but carries around its type. Not making too much use of that, and it would be nice to make it and a column work the same way. (It's called a WherePredicateAtom, which is a bit of a mouthful.)

So, now I have a test passing that looks like this:

`            
List<Columns> found =

                    dao.select(where("integer_column", Operator.GREATER_THAN, 67L)

                            .and("integer_column", Operator.LESS_THAN, 84L)

                            .and(

                                    where("string_column", Operator.LIKE, "%THREE%")

                                    .or("decimal_column", Operator.EQUALS, new BigDecimal("5.0"))

                            ));
`

It could use some clean up, and not all the methods are written for all the types. But I think it's getting pretty close.

**Comment by ojplg** [comment]
Also uses some of your ideas about static imports for the "where" method.

I think it should work easily enough for deletes. Updates are a bit trickier.

**Comment by kazetsukaimiko** [comment]
Ah, you kind of liked that static method idea? where().and.or.and.or, etc. I guess that's the inverse of what I was going for, you're carrying around the type by its value. I was more trying to replace "string column" with `Something<String>` (which has a name "string_column") so that you could enforce specific Operators based on the type of that `Something`, as well as enforce the comparison value's type.

This looks like it works quite well, again without rocking the boat. Well done. I would not however allow to chain where() with or()/and() like you have- what exactly does this mean:

`where(a).and(b).or(c).and(d)`

I would instead do:
`or(and(where(a), where(b)), and(where(c), where(d)));`

THAT is very explicit and clear in what it means. Serial chains and/or are not.

**Comment by kazetsukaimiko** [comment]
As for updates:

```
dao.update(
  template,
  List<String> columns,
  wherePredicateAtom
)
```

Your template model makes a lot more sense here than in selects.

**Comment by ojplg** [comment]
Well, you definitely touched on two things that are bugging me.

1. The possibility of generating SQL that is hard to understand. I have to admit, I do not know what
`select * from table where foo=1 and bar=2 or baz=3` means. I imagine things are just left-associative, but I would have to look it up. It's definitely confusing. There are a bunch of things that could be done. What I did is just one of them. I want to sleep on it and ponder it for a while. If you have some more ideas about this, I would like to hear them.
2. The types of the columns are kind of erased. One thing that's primitive is just say there's a group of numeric operators (=, >, <, etc) and a group of string operators (=, LIKE, NOT LIKE) and since the numeric operators also work on dates, we just need two sets. Then we could just make the string methods take one set, and the other methods take the numeric set. I guess there is one other set for boolean columns. Equals and not equals only. That might be enough. It would be better to share the code between the columns and the where clause generator. And better still to really export the column types and the methods available from them as you suggest. I just don't see a way of doing that. Even if you did something like expose methods like "getBigDecimalColumn(String columnName) on the Dao, the thing is still dependent upon the name being right and describing the correct column type. Otherwise, I don't really have ideas for how to do it.

Those are the two issues that I see. The current code is better than what hrorm has right now, but maybe there is something much better. Should we keep experimenting?

One other thing: we are doing nothing yet to support the IN or NOT IN operators. Those are really tricky to use with prepared statements. Not really possible in fact, you need to just put the values directly into the string, and then you have problems with set size. It could be done of course, but I don't love it.

Are there other operators we are not considering? We're starting to run into realms of SQL that I do not consider part of every day use.

**Comment by kazetsukaimiko** [comment]
> Well, you definitely touched on two things that are bugging me.
> 

For the and/or chaining, this:
`where(a).and(b).or(c).and(d)`
produces:
`select * from table where a AND b OR c AND d`

Again, I don't know what that means! Neither do you, it seems.

Whereas this:
`or(and(where(a), where(b)), and(where(c), where(d)));`

Produces:
`select * from table where ((a AND b) OR (c AND d))`

This grouping is what makes things very clear: a AND b, or , c AND d. And even if d is another composite (e OR f), the grouping removes all ambiguity. This is why serial chaining doesn't make sense unless you do something like:

`or().group().where(a).and(b).endGroup().group().where(c).and(d).endGroup()`

But the nesting model just makes more sense because each nest layer is predictably a (statement) .

> 2. The types of the columns are kind of erased. 

This is the large refactor effort I'd like to do, later. The problem with that is I'm predicting a more or less ground-up rewrite. I think separating the SQL generation + fetching process entirely from the entity hydration layer will give us an overall cleaner, simpler, more easily extensible stack. 

Your recent ideas however are excellent for solving some immediate needs within HRORM itself, as it is.

> One thing that's primitive is just say there's a group of numeric operators (=, >, <, etc) and a group of string operators (=, LIKE, NOT LIKE) and since the numeric operators also work on dates, we just need two sets. Then we could just make the string methods take one set, and the other methods take the numeric set. I guess there is one other set for boolean columns. 

I explored this in a branch already, it isn't perfect, but it works. Extending Operator to StringOperators for instance, and having static methods StringOperators.equals() (overriding Operators.equals()) is enough to make the type system work with opeprator enforcement.

LIKE was a bit harder, but it works as well.

>It would be better to share the code between the columns and the where clause generator.

Again, I think this is a major refactor effort I'd like to undertake if you're up for it.

> And better still to really export the column types and the methods available from them as you suggest. I just don't see a way of doing that. Even if you did something like expose methods like "getBigDecimalColumn(String columnName) on the Dao, the thing is still dependent upon the name being right and describing the correct column type. Otherwise, I don't really have ideas for how to do it.

To "fetch" a field without a static metamodel:
`Fields.bigDecimal(String columnName)`

You'd then rely on RuntimeExceptions to catch problems here. 

But give the option to do static field declarations wherever you wanted to put them, on MyEntity in this case:
```
public static final Field<MyEntity, Long> ID = Fields.define("id", MyEntity::getId, MyEntity::setId);
// All other fields
```

Then when you define the DaoBuilder:

```
new DaoBuilder(/**/)
  .primaryKeyField(MyEntity.ID)
  // All other fields defined the same way
```
This way, you can use the metamodel (and get compile-time safety) or not use the metamodel, and get runtime validation. 

> One other thing: we are doing nothing yet to support the IN or NOT IN operators. Those are really tricky to use with prepared statements. Not really possible in fact, you need to just put the values directly into the string, and then you have problems with set size. It could be done of course, but I don't love it.

I've explored this, and I'm in the same boat as you, ? substitution still works, you just need to keep track of the substitution index.  I think the simplest approach would be to have an object dedicated to building sql that keeps close track of all of the substitution indexes, including IN clauses. 

We might just have to settle for in() excluding other Predicates entirely. I recommend Stream to solve other use cases:
```
List<MyEntity> alotOfMyEntity = dao.in(MyEntity.ID, allTheIds) // Returns Stream<MyEntity>
  .filter(MyEntity::isMarried)
  .filter(myEntity -> myEntity.getAge() < 30)
  .collect(Collectors.toList());
```

Some clever use of Stream could make this page internally, transparently, and especially for NOT IN use cases, returning everything in a fully buffered List is giving the user the ability to OutOfMemory pretty easily. Databases are for more than mere persistence, databases also solve problems with huge datasets that cannot be processed in memory. My first assumption is that anything in an external datasource isn't going to fit into the heap. If there's any room for Streaming, I'd consider it in the next iteration of things.

**Comment by kazetsukaimiko** [comment]
For IN, you can handle upwards of 10,000,000 IDs fairly easily with Stream.

```
Stream<Integer> allTheIds = fetchIds(); // Something providing the Ids

// Creates pages of 1,000 ids and submits them for separate IN queries
dao.in(MyEntity.ID, allTheIds)
  .filter(myPredicate)
  .filter(myOtherPredicate)
  .forEach(myEntity -> doSomething(myEntity));
```

This paging approach doesn't work for NOT IN, obviously. However, I have seen (Oracle) JDBC drivers that disallow > X number of IN comparisons (lets say 1000 in this case) be _very_ lenient with:

```
SELECT * FROM my_entity WHERE 
 ID NOT IN (firstThousand)
 AND
 ID NOT IN (secondThousand)
 AND
 ID NOT IN (thirdThousand)
 ...
 AND
 ID NOT IN (ninthThousand)
```

Very, very, very ugly. But for some reason, its always worked for me.  May be worth testing in H2 to see what the upper limits are.

**Comment by ojplg** [comment]
I have made a lot of internal changes and clean-ups, but functionally my branch is about the same.

I did create a new branch called typed_columns, where I added the type as another generic argument, but after that work, I did not find it helpful. So, I am just stashing it away for now.

It turns out, there's only one line of code per type that gets shared between the where clause and the column. And there's just no way to link up the columns with the where clause except by dynamically loading the column, so you don't get real static type checking anyway.

I am still pondering the syntax for generating where clauses and the issue of having multiple operator types. I am not convinced that one way or another is a big deal.

**Comment by kazetsukaimiko** [comment]
Wow, just went over your branch. There's really a lot there as it is.

> I did create a new branch called typed_columns, where I added the type as another generic argument, but after that work, I did not find it helpful. So, I am just stashing it away for now.

Yeah, I've done the same thing twice already. This is why I concluded that perhaps Column (representing the database) doesn't need a type, but a new Field (representing the entity) type does. 

> It turns out, there's only one line of code per type that gets shared between the where clause and the column. And there's just no way to link up the columns with the where clause except by dynamically loading the column, so you don't get real static type checking anyway.

Depends on what you mean by "real". For example, if I have a private `Map<Field<?>, Object>` that I wrap in an object that enforces typing at insertion, but still does what the compiler considers unchecked casts on the way out:

```
<T> void set(Field<T> field, T thing) {
   map.put(field, thing);
}
@SuppressWarnings("unchecked")
<T> T get(Field<T> field) {
  (T) map.get(field); // Will this ever throw a ClassCastException?
}
```

As long as that map isn't outside accessible, you still effectively have safe typing. At the very least you can give the user compile-time safety in their code.

> I am still pondering the syntax for generating where clauses and the issue of having multiple operator types. I am not convinced that one way or another is a big deal.

Again, the one thing I think should be done is making the user nest with static and()/or(). It looks like you have a way to do that in code but still allow for chaining. If you're fine with the potential for the user to misuse that, its actually a nice convenience. No right or wrong answer here.

For Operators, its easy enough to do, but you've already eliminated the scope creep that was bugging me. 

Gee, I kinda feel bad you're doing all the heavy lifting here. Let me know if there's anything I can help with.

**Comment by kazetsukaimiko** [comment]
And I do want to mention, you've vastly expanded HRORM's capabilities with this Where work.

**Comment by ojplg** [comment]
First of all, you've contributed a lot to this project. For this bit of work, I've done more of the typing, but so what? There's a lot more to moving things forward than that.

I decided to pull together what I had, and do a release, so hrorm 0.7.0 is out.

I am still open to changing up the syntax for building where clauses, since it's debatable. I'm also open to making subtypes of the Operator class, to try to cut down on errors. But I was too lazy right now.

I just ripped out the Map<String, Operator> and the Operator.ranges stuff, since that was pretty clunky, and now we have something better.

I agree that hrorm is much better after this work, even though it could probably use more refinement and testing.

**Comment by ojplg** [comment]
So, we have a pretty powerful and flexible facility for constructing arbitrary where clauses with OR and AND and nesting. I am not sure the syntax and types are the best. However the main issues raised by this ticket are complete.

If there are specific issues with either functionality or improvements to the style or safety of the components, I would like new tickets for that.

In sum, I think this ticket has been completed.
